Assets/MRTK/Providers/LeapMotion/LeapMotionArticulatedHand.cs

```diff
@@ -0,0 +1,299 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.﻿
+
+using System.Collections.Generic;
+using UnityEngine;
+using Microsoft.MixedReality.Toolkit.Input;
+using Microsoft.MixedReality.Toolkit.Utilities;
+using System;
+
+#if LEAPMOTIONCORE_PRESENT
+using Leap.Unity.Attachments;
+using Leap.Unity;
+using Leap;
+#endif
+
+namespace Microsoft.MixedReality.Toolkit.LeapMotion.Input
+{
+
+    [MixedRealityController(
+    SupportedControllerType.ArticulatedHand,
+    new[] { Handedness.Left, Handedness.Right })]
+    /// <summary>
+    /// Class that represents one Leap Motion Articulated Hand.
+    /// </summary>
+    public class LeapMotionArticulatedHand : BaseHand
+    {
+        /// <summary>
+        /// Constructor for a Leap Motion Articulated Hand
+        /// </summary>
+        /// <param name="trackingState">Tracking state for the controller</param>
+        /// <param name="controllerHandedness">Handedness of this controller (Left or Right)</param>
+        /// <param name="inputSource">The origin of user input for this controller</param>
+        /// <param name="interactions">The controller interaction map between physical inputs and the logical representation in MRTK</param>
+        public LeapMotionArticulatedHand(TrackingState trackingState, Handedness controllerHandedness, IMixedRealityInputSource inputSource = null, MixedRealityInteractionMapping[] interactions = null) 
+            : base(trackingState, controllerHandedness, inputSource, interactions)
+        {
+            handDefinition = new ArticulatedHandDefinition(inputSource, controllerHandedness);
+#if LEAPMOTIONCORE_PRESENT
+            SetAttachmentHands();
+#endif
+        }
+
+        private readonly ArticulatedHandDefinition handDefinition;
+
+        // Set the interations for each hand to the Default interactions of the hand definition
+        public override MixedRealityInteractionMapping[] DefaultInteractions => handDefinition?.DefaultInteractions;
+
+#if LEAPMOTIONCORE_PRESENT
+
+        protected LeapMotionDeviceManager LeapMotionDeviceManager => CoreServices.GetInputSystemDataProvider<LeapMotionDeviceManager>();
```

Comment:
@keveleigh @davidkline-ms as a more general design question, it's interesting to me that a provider's internal code has to reach outside of the provider to get it - i.e. this is a hand created by a provider, and in order to get the provider itself, it has to reach outside of its typical bounds.

Reply:
Yeah, I agree this pattern feels odd, though I think it's the only way currently for a controller to know its device manager.

In this case, it looks like this is mainly used to grab the corresponding Leap `AttachmentHand` from the device manager and cache it. @CDiaz-MS Do you think this might be place (like with `Hand`) that the constructor can take these in as parameters or can be set manually by the data provider? Maybe with a `SetHands(AttachmentHand attachmentHand, Hand hand)` method or something, if we don't want to mess with the constructor's parameters and add plug-in specific types to it. It feels like these are things that the data provider can actively provide to the MRTK class instead of needing to be asked for it.

Reply:
Adding leap specific types to the constructor is possible, but I would have to ifdef them out if the core assets are not in the project because they would be invalid.   It doesn't feel right modifying the constructor in this way and might lead to other issues.

The only elements I need from the device manager are the `AttachmentHands` and `LeapServiceProvider`.  The `AttachmentHands` and the `LeapServiceProvider` are added to the scene at runtime, I could do a `GameObject.Find`.  @keveleigh What do you think?

Reply:
Agreed 100% on not changing the constructor. Good points!
What about something like an `internal void SetHands(AttachmentHand attachmentHand, Hand hand)` method that the device manager can call immediately after creating the hand? It looks like the manager keeps track of `AttachmentHand`s and `Hand`s based on handedness, so it should be able to tell what maps to what. Then the leap hand can use its locally cached references at runtime.

I'd recommend against `GameObject.Find` any time we can get around using it. Always a cost to searching the whole scene for something!